AZSLc Crash On Shader Option With Undefined Type

galibzon opened the pull request:
This is version 1.7.30

A clear error is now emitted when a shader option is of undefined type.

Previously it was only asserting and the user was NOT provided with
a message describing the root cause of the problem.

Signed-off-by: garrieta <garrieta@amazon.com>

Commit:
AZSLc Crash On Shader Option With Undefined Type

This is version 1.7.30

A clear error is now emitted when a shader option is of undefined type.

Previously it was only asserting and the user was NOT provided with
a message describing the root cause of the problem.

Signed-off-by: garrieta <garrieta@amazon.com>

santorac commented on src/AzslcSemanticOrchestrator.cpp:
```diff
@@ -567,6 +567,12 @@ namespace AZ::ShaderCompiler
+            if (!varInfo.m_typeInfoExt.IsClassFound())
+            {
+                const string message = FormatString("Unknown type '%s' for shader option '%s'",
+                    varInfo.m_typeInfoExt.GetDisplayName().c_str(), uqNameView.data());
```
If uqNameView is a string_view, you need to use '%.*s' and specify the length as an argument.

santorac replied:
(if it happens to be a string, use c_str() instead of data())

galibzon replied:
uqNameView is of type UnqualifiedNameView, which is a subclass of 'string_view'.
Although 'string_view' "can be evil", in particular when you don't know where it comes from, e.g: A library, etc.
It is safe to use "UnqualifiedNameView.data()" throughout AZSLc  because they are initialized always from null terminated strings.

galibzon replied:
To add more context...

UnqualifiedNameView is used heavily throughout AZSLc, and because of that you'll see many uses of ".data()".
Your request for change, to make an impact, would have to be applied throughout the code base.

And I'd be more than happy to do as many changes as necessary If I knew there were string_views in the code that came from somewhere else.

Commit:
Safe usage of string_view.data() during string formatting.

Instead of
"%s", string_view.data()
Changed to
"%.*s", string_view.size(), string_view.data()

Signed-off-by: garrieta <garrieta@amazon.com>

Commit:
static casted the string_view.size() to 'int' for correcteness.

Signed-off-by: garrieta <garrieta@amazon.com>
